src/Lean/Meta/Tactic/Rewrite.lean

```diff
@@ -40,13 +40,13 @@ def _root_.Lean.MVarId.rewrite (mvarId : MVarId) (e : Expr) (heq : Expr)
           -- construct rewrite proof
           let eNew := eAbst.instantiate1 rhs
           let eNew ← instantiateMVars eNew
-          let eEqE ← mkEq e e
-          let eEqEAbst := mkApp eEqE.appFn! eAbst
-          let motive := Lean.mkLambda `_a BinderInfo.default α eEqEAbst
+          let eType ← inferType e
+          let motive := Lean.mkLambda `_a BinderInfo.default α eAbst
           unless (← isTypeCorrect motive) do
             throwTacticEx `rewrite mvarId "motive is not type correct"
-          let eqRefl ← mkEqRefl e
-          let eqPrf ← mkEqNDRec motive eqRefl heq
+          let u1 ← getLevel α
+          let u2 ← getLevel eType
+          let eqPrf ← instantiateMVars (mkApp6 (.const ``congrArg [u1, u2]) α eType lhs rhs motive heq)
```

Comment:
```suggestion
          let eqPrf := mkApp6 (.const ``congrArg [u1, u2]) α eType lhs rhs motive heq
```
Let’s see if this is needed, and maybe the cause for the reordering of side-condition goals.